ThreeDSecure/src/main/java/com/braintreepayments/api/threedsecure/ThreeDSecureClient.kt

```diff
@@ -0,0 +1,426 @@
+ * for a full explanation of 3D Secure.
+ */
+@Suppress("TooManyFunctions")
+class ThreeDSecureClient internal constructor(
+    private val braintreeClient: BraintreeClient,
+    private val cardinalClient: CardinalClient = CardinalClient(),
+    private val api: ThreeDSecureAPI = ThreeDSecureAPI(braintreeClient)
+) {
+    /**
+     * Initializes a new [ThreeDSecureClient] instance
+     *
+     * @param context an Android Context
+     * @param authorization a Tokenization Key or Client Token used to authenticate
+     */
+    constructor(
+        context: Context,
+        authorization: String
+    ) : this(BraintreeClient(context, authorization))
+
+    /**
+     * Call this method to initiate the 3D Secure flow.
+     *
+     * Verification is associated with a transaction amount and your merchant account. To specify a
+     * different merchant account (or, in turn, currency), you will need to specify the merchant
+     * account id when [
+     * generating a client token](https://developer.paypal.com/braintree/docs/start/hello-client#get-a-client-token)
+     *
+     * During lookup the original payment method nonce is consumed and a new one is returned, which
+     * points to the original payment method, as well as the 3D Secure verification. Transactions
+     * created with this nonce will be 3D Secure, and benefit from the appropriate liability shift
+     * if authentication is successful or fail with a 3D Secure failure.
+     *
+     * @param context  Android context
+     * @param request  the [ThreeDSecureRequest] with information used for authentication.
+     * @param callback [ThreeDSecureResultCallback]
+     */
+    fun createPaymentAuthRequest(
+        context: Context,
+        request: ThreeDSecureRequest,
+        callback: ThreeDSecurePaymentAuthRequestCallback
+    ) {
+        braintreeClient.sendAnalyticsEvent(ThreeDSecureAnalytics.VERIFY_STARTED)
+        if (request.amount == null || request.nonce == null) {
+            callbackCreatePaymentAuthFailure(
+                callback,
+                ThreeDSecurePaymentAuthRequest.Failure(
+                    InvalidArgumentException(
+                        "The ThreeDSecureRequest nonce and amount cannot be null"
+                    )
+                )
+            )
+            return
+        }
+
+        braintreeClient.getConfiguration { configuration: Configuration?, error: Exception? ->
+            val failure = when {
+                configuration == null -> {
+                    error ?: BraintreeException("Configuration is null")
+                }
+
+                !configuration.isThreeDSecureEnabled -> {
+                    BraintreeException(
+                        "Three D Secure is not enabled for this account. " +
+                            "Please contact Braintree Support for assistance."
+                    )
```

Comment:
```suggestion
                    val message = "Three D Secure is not enabled for this account. Please contact Braintree Support for assistance."
                    BraintreeException(message)
```
suuuper nit: It isn't much of a performance boost since it's not frequently called in a loop or anything, but string concatenation does have a runtime cost. The linter may be influencing the implementation here as well.

Also definitely out of scope but I wonder if we should consider externalizing our error strings into XML in the future? We may be able to inject some documentation links to help merchants who get stuck.

Reply:
This is a great callout. I agree the string concatenation here is neglible, but we should find a long term solution for handling longer strings. Maybe some sort of StringLoader class that either grabs strings from XML or has them defined internally.